### PR TITLE
Made VersionChecker and Jenkins stuff available as mini-APIs

### PR DESCRIPTION
**Pull request by kantenkugel**
Overview
======

This PR introduces following changes:
- `JenkinsApi` is now no longer static and bound to JDA. This allows for retrieval of builds for different projects (like Yui for example)
- The previous `VersionedItem` was completely revamped to allow other elements to hook into features that were only available through hardcoded code:
  - Added easy way to handle updates through update handler
  - Added possibility to provide Changelogs through `VersionedItem` (used in `!changelog` command)
    - JDA itself now uses this instead of hardcoded command
    - Added utility `JenkinsChangelogProvider` class to fetch changelogs from Jenkins builds (what JDA uses)
  - Added possibility to have custom version-retrieval code instead of relying purely on maven repos
    - The utility class `JenkinsVersionSupplier` uses artifact name / build number of Jenkins to do this
  - The `!version` / `!latest` command now shows JDA, Lavaplayer and JDA-Utilities if no argument is provided, otherwise it gets the `VersionedItem`(s) by name / alias
  - A lot of small changes like ability for aliases for items
- Added 2 new items to be tracked
  - JDAction (version tracking) - Gradle plugin which aborts the build if there are unused RestActions in the code
  - Yui (version + changelog) - Example bot written by DV8FromTheWorld
- Moved some code to better reflect current situation and for cleanup

Short Usage Help
===========

To use the new System, its probably best to just look at how items are registered in `VersionCheckerRegistry#register()` and looking at a few `VersionedItem`s like `JDAItem`.

Alternatively, here is a short summary of what to do:

- Create a new `SimpleVersionedItem` instance if you don't need advanced features and only need to track a maven repo for versions, otherwise create a new class extending `VersionedItem`
- Register the item in `VersionCheckerRegistry#register()`
- If you need custom version-retrieval code, override `VersionItem#getCustomVersionSupplier()`
  - If you want to get the version from Jenkins, you can use `JenkinsVersionSupplier`
  - For example code, look at `JenkinsVersionSupplier` or `JDActionItem`
- To use the on-update hook, override `VersionedItem#getUpdateHandler()`
  - Example is found in `JDAItem`
- If you want to integrate into the `!changelog` command, override `VersionedItem#getChangelogProvider()`
  - For Jenkins as changelog provider (like JDA uses it), you can use `JenkinsChangelogProvider`

TODO
====

- [x] Find more stuff for todo list
- [x] Add docs to `VersionedItem` and other important parts
- [x] Add implementation help to PR
- [x] Change all the maven/gradle commands to be able to handle all versioned items that rely on maven repos for versioning
- [x] Review of full PR

### CODE REVIEW

**Review by Almighty-Alpaca — APPROVED** [review]

**Comment by kantenkugel** [comment]
Just added my last feature commit to this. Going to review the whole PR again today before marking it as complete

File: src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/MavenProjectCommand.java [hunk 1]
```diff
@@ -34,31 +33,22 @@
...
-
-        final StringBuilder builder = new StringBuilder();
+        VersionedItem jdaItem = VersionCheckerRegistry.getItem("jda");
+        LinkedList<VersionedItem> items = VersionCheckerRegistry.getItemsFromString(content).stream()
+                //only allow items which use maven for versioning
+                .filter(item -> item.getCustomVersionSupplier() == null)
+                .collect(Collectors.toCollection(LinkedList::new));
+        //force jda to be at first position
+        items.remove(jdaItem);
+        items.addFirst(jdaItem);
```
Comment (kantenkugel) [inline src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/MavenProjectCommand.java]: Moving these blocks to some utility class (duplicate code)

File: src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/VersionsCommand.java [hunk 2]
```diff
@@ -23,19 +25,38 @@ public void dispatch(final User sender, final TextChannel channel, final Message
...
-        eb.setTitle(EmbedBuilder.ZERO_WIDTH_SPACE, null);
+        List<VersionedItem> items;
+        if(content.trim().isEmpty())
+        {
+            items = VersionCheckerRegistry.getItemsFromString(DEFAULT_ITEMS);
+        }
+        else
+        {
+            items = VersionCheckerRegistry.getItemsFromString(content);
+        }
```
Comment (kantenkugel) [inline src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/VersionsCommand.java]: probably better to just drop the braces here

File: src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/VersionsCommand.java [hunk 3]
```diff
@@ -23,19 +25,38 @@ public void dispatch(final User sender, final TextChannel channel, final Message
...
             {
-                eb.addField(versionedItem.getName(), versionedItem.getVersion(), true);
+                if (versionedItem.getUrl() != null)
+                {
+                    eb.addField(versionedItem.getName(), String.format("[%s](%s)", versionedItem.getVersion(), versionedItem.getUrl()), true);
+                }
+                else
+                {
+                    eb.addField(versionedItem.getName(), versionedItem.getVersion(), true);
+                }
```
Comment (kantenkugel) [inline src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/VersionsCommand.java]: Maybe drop braces here too

File: src/main/java/com/kantenkugel/discordbot/versioncheck/changelog/JenkinsChangelogProvider.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                changes = Collections.singletonList("No git commits assigned");
+            }
+
+            changelogs.add(new Changelog(title, changes, build.getUrl()));
+        }
+        return changelogs;
+    }
+
+    private List<String> getChangelog(List<JenkinsChange> changeSet)
+    {
```
Comment (kantenkugel) [inline src/main/java/com/kantenkugel/discordbot/versioncheck/changelog/JenkinsChangelogProvider.java:84]: could probably be made static

**Review by kantenkugel — COMMENTED** [review]
Some stuff i found

File: src/main/java/com/kantenkugel/discordbot/versioncheck/changelog/JenkinsChangelogProvider.java (hunk 4 above)
Comment (kantenkugel) [reply src/main/java/com/kantenkugel/discordbot/versioncheck/changelog/JenkinsChangelogProvider.java:84]: Requires the instance variable githubUrl, so no static avail without extra param

**Comment by kantenkugel** [comment]
Everything ready... requesting final review and merge from @Almighty-Alpaca

**Review by Almighty-Alpaca — APPROVED** [review]
Looks good to me 👍

**Pull request merged**